opsdroid/database/matrix/__init__.py

```diff
@@ -0,0 +1,260 @@
+        return (
+            self.connector._allow_encryption
+            and self.room_id in self.connector.connection.store.load_encrypted_rooms()
+        )
+
+    async def migrate_database(self):
+        """Migrate existing 'opsdroid.database' state event to 'dev.opsdroid.database' event."""
+
+        data = await self.connector.connection.room_get_state(room_id=self.room_id)
+        if isinstance(data, RoomGetStateError):
+            _LOGGER.error(
+                f"Error migrating from opsdroid.database to {self._event_type} in room {self.room_id}: {data.message}({data.status_code})"
+            )
+            return
+
+        if not any(event["type"] == self._event_type for event in data.events):
+            for event in data.events:
+                if event["type"] == "opsdroid.database" and event["content"]:
+                    await self.connector.connection.room_put_state(
+                        room_id=self.room_id,
+                        event_type=self._event_type,
+                        content=event["content"],
+                        state_key=event["state_key"],
+                    )
+                    await self.connector.connection.room_redact(
+                        room_id=self.room_id, event_id=event["event_id"],
+                    )
+
+        self.should_migrate = False
+
+    async def connect(self):
+        """Connect to the database."""
+
+        _LOGGER.info("Matrix Database connector initialised.")
+
+    async def put(self, key, value):
+        """Insert or replace a value into the database for a given key."""
+
+        if self.should_migrate:
+            await self.migrate_database()
+
+        # If the single state key flag is set then use that else use state key.
+        state_key = (
+            "" if self._single_state_key is True else self._single_state_key or key
+        )
+
+        if self._single_state_key:
+            value = {key: value}
+        elif not isinstance(value, dict):
+            raise ValueError(
+                "When the matrix database is configured with single_state_key=False, "
+                "the value passed must be a dict."
+            )
+
+        data = await self.get(state_key, check_dict=value)
+        if data is None:  # dict could be empty
+            _LOGGER.error("Error putting key into matrix room")
+            return
+
+        if {**data, **value} == data:
+            _LOGGER.error("Not updating matrix state, as content hasn't changed.")
+            return
+        else:
+            data = {**data, **value}
+
+        if self.is_room_encrypted and self.should_encrypt:
+            for k, v in value.items():
+                room_event = await self.opsdroid.send(
+                    GenericMatrixRoomEvent(
+                        target=self.room_id, content={k: v}, event_type=self._event_type
+                    )
+                )
+                data[k] = {"encrypted_val": room_event.event_id}
```

Comment:
When I said in matrix that we should only send one event per put what I think I meant is this:
```suggestion
            room_event = await self.opsdroid.send(
                GenericMatrixRoomEvent(
                    target=self.room_id, content=value, event_type=self._event_type
                )
            )
            data[k] = {"encrypted_val": room_event.event_id}
```